### PR TITLE
Adding trimming sku value function to sku backend model.

### PR DESCRIPTION
**Pull request by bartoszkubicki**
### Description
I have added function to sku backend model which trimms value from whitespace.

### Fixed Issues (if relevant)

1. https://github.com/magento/magento2/issues/16572 Trim whitespace on SKU when saving product
2. https://github.com/magento/magento2/issues/12300 SKU values are not trimmed with the space.

### Manual testing scenarios

1. Edit sku, by adding some whitespace
2. Save product.
3. Sku will be trimmed and saved in changed form.
4. Messsage that sku has been changed will be displayed.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @bartoszkubicki. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by ihor-sviziev** [comment]
Hi @bartoszkubicki,
We already have PR for fixing the same issue and it looks absolutely different. https://github.com/magento/magento2/pull/16592

@VladimirZaets could you check this PR?

**Comment by bartoszkubicki** [comment]
@ihor-sviziev Actuallly haven't noticed that. I looked at this PR, but I am not sure - adding `trim` function everywhere seems to be quite redundant, in my opinion backend model is proper place for validation/modyfing attribute value. It should cover most of use-cases, but actually I can't make sure it covers all.

**Comment by rogyar** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rogyar. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rogyar, here is your [Magento instance](https://i-18019-2-3-develop.engcom.dev.magento.com).
Admin access: [https://i-18019-2-3-develop.engcom.dev.magento.com/admin](https://i-18019-2-3-develop.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by rogyar** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @rogyar. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @rogyar, here is your [new Magento instance](https://pr-18019.engcom.dev.magento.com).
Admin access: [https://pr-18019.engcom.dev.magento.com/admin](https://pr-18019.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

**Comment by rogyar** [comment]
I found this solution more universal than in #16592
The only issue I found is when we create a configurable product with whitespace in SKU and generate child products (before the parent product is saved), the whitespace will be included in the child products SKU. However, the main issue is fixed

File: app/code/Magento/Catalog/Model/Product/Attribute/Backend/Sku.php [hunk 1]
```diff
@@ -127,4 +128,18 @@ protected function _getLastSimilarAttributeValueIncrement($attribute, $object)
         $data = $connection->fetchOne($select, $bind);
         return abs((int)str_replace($value, '', $data));
     }
+
```
Comment (rogyar) [inline app/code/Magento/Catalog/Model/Product/Attribute/Backend/Sku.php:131]: Please, remove this extra line

**Review by rogyar — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/Product/Attribute/Backend/Sku.php (hunk 1 above)
Comment (bartoszkubicki) [reply app/code/Magento/Catalog/Model/Product/Attribute/Backend/Sku.php:131]: Done.

**Comment by bartoszkubicki** [comment]
I have also include fix for not resolved case - I have added custom frontend validation js rules and included it for sku field by field ui modifier.

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-2998](https://jira.corp.magento.com/browse/ENGCOM-2998) has been created to process this Pull Request

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-2998](https://jira.corp.magento.com/browse/ENGCOM-2998) has been created to process this Pull Request

**Comment by bartoszkubicki** [comment]
@magento-engcom-team Is it possible to add this fix to earlier Magento versions? Does it require my action or you can do it internally?

**Comment by magento-engcom-team** [comment]
Hi @bartoszkubicki. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.

**Pull request merged**